server/broadcast.go

```diff
@@ -473,34 +475,55 @@ func processSegment(cxn *rtmpConnection, seg *stream.HLSSegment) ([]string, erro
 		sv = verification.NewSegmentVerifier(Policy)
 	}
 
-	for i := 0; i < MaxAttempts; i++ {
-		// if fails, retry; rudimentary
-		var urls []string
-		if urls, err = transcodeSegment(cxn, seg, name, sv); err == nil {
-			return urls, nil
+	var (
+		startTime = time.Now()
+		attempts  []data.TranscodeAttemptInfo
+		urls      []string
+	)
+	for len(attempts) < MaxAttempts {
+		// if transcodeSegment fails, retry; rudimentary
+		var info data.TranscodeAttemptInfo
+		urls, info, err = transcodeSegment(cxn, seg, name, sv)
+		attempts = append(attempts, info)
+		if err == nil {
+			break
 		}
 
 		if shouldStopStream(err) {
 			glog.Warningf("Stopping current stream due to err=%v", err)
 			rtmpStrm.Close()
-			return nil, err
+			break
 		}
-
 		if isNonRetryableError(err) {
 			glog.Warningf("Not retrying current segment nonce=%d seqNo=%d due to non-retryable error err=%v", nonce, seg.SeqNo, err)
-			return nil, err
+			break
 		}
-
 		// recoverable error, retry
 	}
-	if err != nil {
+	if MetadataQueue != nil {
+		var (
+			key      = fmt.Sprintf("stream_health.transcode.%c.%s", mid[0], mid)
+			segMeta  = data.SegmentMetadata{seg.Name, seg.SeqNo, seg.Duration, len(seg.Data)}
+			success  = err == nil && len(urls) > 0
+			transEvt = data.NewTranscodeEvent(monitor.NodeID, string(mid), segMeta, startTime, success, attempts)
+		)
+		BackgroundPublish(MetadataQueue, key, transEvt)
+	}
+	if len(attempts) == MaxAttempts && err != nil {
 		err = fmt.Errorf("Hit max transcode attempts: %w", err)
 	}
-	return nil, err
+	return urls, err
 }
 
 func transcodeSegment(cxn *rtmpConnection, seg *stream.HLSSegment, name string,
-	verifier *verification.SegmentVerifier) ([]string, error) {
+	verifier *verification.SegmentVerifier) (urls []string, info data.TranscodeAttemptInfo, err error) {
+	defer func(startTime time.Time) {
+		info.LatencyMs = time.Since(startTime).Milliseconds()
+		if err != nil {
+			errStr := err.Error()
+			info.Error = &errStr
```

Comment:
Could we just store the error itself to preserve that information for as long as possible in case its needed later on as opposed to only storing the string representation here?

Reply:
The annoying thing is that the `error` is not json serializable, so I'd need to create a custom type, either a struct with all the same fields but `error`s instead of `string`s, and convert them to `string`s in the last minute, or like an `error` wrapper that holds an `error` inside but provides custom `MarshalJSON` functions to serialize them nicely (like this [`UnixNanoTime`](https://github.com/livepeer/livepeer-data/blob/9b508371591e5679d54e01d8289dd78916b12ad6/pkg/data/time.go#L8))

Do you think either of those are worth it?

Reply:
Got it. Keeping it as a string seems reasonable then. Although curious why that field is a [string pointer](https://github.com/livepeer/livepeer-data/blob/d24bb6e4e42d5407041f6839e6ce6d1edc0dae42/health/events.go#L54) as opposed to string given that the other string fields in structs from that package are not pointers?

Reply:
So the `*string` was more of a "stylistic" choice for the JSON output object, so that the event object for a successful attempt could explicitly have an `"error": null` field. That seemed better than the `"error": ""` or no `"error"` field at all (if we used `omitempty`) in the JSON. That especially because, in the end, this object is a "user-facing" JSON that will be parsed from any language, not only `go`, like for example on the browser via SSE. So then:
 - `"error": ""`: This would fail a simple check like `error != null` on a language where strings are nullable
 - `omitempty`: I thought it could be confusing to not have any `"error"` field in most events (assuming most are successful), like a bit harder to validate the schema when building something on top of it without "trusting" that the field will be there in case an error happened. (although just `null` doesn't 100% guarantee it will be a string either, but at least the field name is there)
 
Anyway, as on the first time, if you dislike the `*string` I can change it as well, not a strong preference here. If I create that "error wrapper" that I proposed, I could have it serialize as `null` when no `error` is there, or just make it a pointer as well which would still be less weird since it'd not be a `string`. I've kind of accustomed to using a pointer to a literal when the value is optional so I find it ok, but it might not be coherent with the `go-livepeer` style.

Reply:
@yondonfu When you get back to this PR, just a reminder of this thread. Let me know if you'd prefer to have the `ErrorString` type on `livepeer-data/pkg/data` or if you have any other suggestions.

Reply:
I don't dislike the use of `*string` - I mainly wanted to understand why we were using a pointer instead of a value so thanks for explaining. I'm cool with keeping the `*string` type for the error.